deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/iterators/VirtualIterator.java

```diff
@@ -0,0 +1,63 @@
+package org.deeplearning4j.spark.parameterserver.iterators;
+
+import javafx.beans.InvalidationListener;
+import javafx.beans.Observable;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+import java.util.function.Consumer;
+
+/**
+ * This class is thin wrapper, to provide block-until-depleted functionality in multi-threaded environment
+ *
+ * @author raver119@gmail.com
+ */
+@Slf4j
+public class VirtualIterator<E> extends java.util.Observable implements Iterator<E> {
+    // TODO: use AsyncIterator here?
+    protected Iterator<E> iterator;
+    protected AtomicBoolean state = new AtomicBoolean(true);
+
+    public VirtualIterator(@NonNull Iterator<E> iterator) {
+        this.iterator = iterator;
+    }
+
+
+    @Override
+    public boolean hasNext() {
+        boolean u = iterator.hasNext();
+        state.compareAndSet(true, u);
+        if (!state.get()) {
+            this.setChanged();
+            notifyObservers();
+        }
+        return u;
+    }
+
+    @Override
+    public E next() {
+        return iterator.next();
```

Comment:
P.s. and if someone breaks it (who?), he'll get NoSuchElementException from underlying iterator.